### ISSUE TITLE
Image format error

### ISSUE THREAD

**Issue by emresaracoglu** [samayo/bulletproof#132]
Hello,

Thank you for your repo. 
I can upload images using Uppy, but I want to use your library in the background.

```
if ($image["files"]) {
	$upload = $image->setSize(50000, 5242880)->setMime(array("jpg", "png", "gif", "bmp", "JPG", "PNG", "GIF", "BMP", "mp4", "MP4"))->setDimension(400, 2500)->setLocation('/var/www/clients/client7/web31/web/')->upload();
	if ($upload) {
		echo $upload->getFullPath(); // uploads/cat.gif
	} else {
		echo $image->getError();
	}
```

I am getting the following error with the above code.

```
<br />
<b>Warning</b>:  Illegal offset type in <b>/var/www/clients/client7/web31/web/vendor/samayo/bulletproof/src/bulletproof.php</b> on line <b>149</b><br />
Invalid File! Only (jpg, png, gif, bmp, JPG, PNG, GIF, BMP, mp4, MP4) image types are allowed
```

Why is this issue occurring even though I have allowed the image format?
Thank  you

**Comment by emresaracoglu** [comment]
I did not make any changes to the source code. These are the codes;

HTML:
```
<form id="niyp">
<div id="newItemPhotoUploadArea"></div>
</div>
```

PHP:
```
<?php
	global $pathWebHost, $db, $vtItems, $image;
	$itemID = isset($_POST['itemID']) ? $_POST['itemID'] : null;
	$target = '/var/www/clients/client7/web31/web/';
	header('Content-Type: application/json'); 
		if ($image["files"]) {
			$upload = $image->setSize(50000, 5242880)->setMime(array("jpg", "png", "gif", "bmp", "JPG", "PNG", "GIF", "BMP", "mp4", "MP4"))->setDimension(400, 2500)->setLocation('/var/www/clients/client7/web31/web/')->upload();
			if ($upload) {
				echo $upload->getFullPath(); // uploads/cat.gif
			} else {
				echo $image->getError();
			} 
	}
```

JS for Uppy
```
const uppy = Uppy.Core({debug: false, autoProceed: true})
			.use(Uppy.Dashboard, {
				target: '#niyp',
				hideAfterFinish: false,
				inline: true,
				disableInformer: false,
				hideUploadButton: true
			})
			.use(Uppy.Form, {target: '#newItemForm'})
			.use(Uppy.XHRUpload, {endpoint: 'https://example.com/samayoUploader'})
			.run()
		uppy.on('success', (fileCount) => {
			console.log(`${fileCount} files uploaded`)
		})
		uppy.on('upload', (data) => {
			console.log(data);
		})
		uppy.on('complete', (result) => {
			console.log('successful files:', result.successful)
			console.log('failed files:', result.failed)
		})
```

**Comment by samayo** [comment]
Hi, you have ```if ($image["files"])``` in your code but I don't see $image = new Bulletproof(...) In your code.

Can you use the example provided in the source code? Both the html and PHP as it is shown? That will tell me if the issue is with the library or somewhere in your code

**Comment by samayo** [comment]
There is nothing wrong I see in your code, can you please try to upload only using the example shown in the readme? If that shows the same error the I will try to check the library. It is not meant to show, the same error as you have . But if you get the same error the we can definitely say the issue is with the library

**Comment by emresaracoglu** [comment]
I tested it in a completely independent area from the framework. To be sure of everything.
In this test, I used both Uppy and the sample code in the README.me file.

HTML:

```
<!doctype html>
<html lang="en">
<head>
	<meta name="author" content="">
	<title>Sticky Footer Template for Bootstrap</title>
	<link href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css" rel="stylesheet">
	<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/uppy@0.24.4/dist/uppy.css">
</head>
<body>
<main role="main" class="container">
	<span>test</span>
	<div id="newItemPhotoUploadArea"></div>
	<form method="POST" enctype="multipart/form-data" action="https://recabin.com/wos/itemPhotoUploader">
		<input type="hidden" name="MAX_FILE_SIZE" value="1000000"/>
		<input type="file" name="pictures" accept="image/*"/>
		<input type="submit" value="upload"/>
	</form>
</main>
<script src="https://code.jquery.com/jquery-3.6.0.min.js"
        integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script>
<script src="https://recabin.com/wos/staticJs"></script>
<script src="https://assets.recabin.com/frontend/ho/vendor/bootstrap/dist/js/bootstrap.bundle.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/uppy@0.24.4/dist/uppy.js"></script>
</body>
</html>
```

upload.php (I used `pictures` for simple form and `files` for Uppy.)
```
include "vendor/autoload.php";

$image = new Bulletproof\Image($_FILES);
$image->setLocation('/var/www/clients/client7/web31/web/', 0777);

if($image["pictures"]){
  $upload = $image->upload(); 

  if($upload){
    echo $upload->getFullPath();  
  }else{
    echo $image->getError(); 
  }
}
```

I tested it in a different area than the framework. I did not encounter any issues. Uppy could not load due to an array problem and I fixed it from Uppy codes.

However, apart from Uppy, I also used the simple form code in the README.me file. I got the same error on both.

If I use your code like this, I keep getting this error:

`$upload = $image->setSize(50000, 5242880)->setMime(array("jpg", "png", "gif", "bmp", "JPG", "PNG", "GIF", "BMP", "mp4", "MP4"))->setDimension(2500, 2500)->setLocation('/var/www/clients/client7/web31/web/')->upload();`

`Invalid File! Only (jpg, png, gif, bmp, JPG, PNG, GIF, BMP, mp4, MP4) image types are allowed`

If I remove `setMime` there is no problem. Could image format `Array` be a problem?

**Comment by samayo** [comment]
Thanks for your time, I think I understand now the issue. There are two things that are producing the two errors (PHP warning and the Invalid File errors).

The first thing is that you are trying to upload file types not supported by the library, the supported list is

`
protected $acceptedMimes = array(
      1 => 'gif', 'jpeg', 'png', 'swf', 'psd',
      'bmp', 'tiff', 'tiff', 'jpc', 'jp2', 'jpx',
      'jb2', 'swc', 'iff', 'wbmp', 'xbm', 'ico'
    );
`

If you try to upload unsupported file type e.g., MP3, then you will get the Invalid File error, another thing to avoid is method chaining such as foo()->bar() because the first method (in your case the mime validator) reruns false the second will trigger error, so it safe to only chain methods that do no involve user input.


So to recap, only upload allowed image type and avoid method chaining.

One last thing some images may have corrupted headers because they were edited or decoded wrongly and they may be identified as non valid image types so if one jpg does not work try the rest and they will work for you.


Let me know if you have more questions

**Comment by emresaracoglu** [comment]
I will make the necessary changes. I wanted to use 'Bulletproof' and I'm glad it's not an issue anymore. 

I sincerely **thank you** for your help.

**Comment by samayo** [comment]
Glad to help, I will close the issue now but feel free to ask for more if have any question
